src/parse.rs

```diff
@@ -290,6 +295,64 @@ impl<'input, 'callback> Parser<'input, 'callback> {
                     }
                     self.tree[cur_ix].item.body = ItemBody::Text;
                 }
+                ItemBody::MaybeMath(is_display, can_open, _can_close, brace_context) => {
+                    let mut search_count = if is_display { 2 } else { 1 };
+                    if !can_open {
+                        search_count -= 1;
+                        if search_count == 0 {
+                            self.tree[cur_ix].item.body = ItemBody::Text;
+                            prev = cur;
+                            cur = self.tree[cur_ix].next;
+                            continue;
+                        }
+                    }
+                    let is_display = search_count > 1;
+
+                    if math_delims.is_populated() {
+                        // we have previously scanned all math environment delimiters,
+                        // so we can reuse that work
+                        if let Some((scan_ix, true)) = math_delims.find(cur_ix, search_count, brace_context, is_display) {
+                            self.make_math_span(cur_ix, scan_ix);
+                        } else {
+                            self.tree[cur_ix].item.body = ItemBody::Text;
+                        }
+                    } else {
+                        // we haven't previously scanned all codeblock delimiters,
+                        // so walk the AST
+                        let mut scan;
+                        'search_count: loop {
```

Comment:
The second time this loops around, I think `math_delims` should already be used.  Otherwise this has quadratic time complexity in a case like `$$x$}$$x$}$$x$}...`

```
$ python -c 'print("$$x$}"*10000)' | time target/release/pulldown-cmark -M > /dev/null
7.53user 0.00system 0:07.62elapsed 98%CPU (0avgtext+0avgdata 6100maxresident)k
0inputs+0outputs (0major+1291minor)pagefaults 0swaps
$ python -c 'print("$$x$}"*20000)' | time target/release/pulldown-cmark -M > /dev/null
31.44user 0.01system 0:31.56elapsed 99%CPU (0avgtext+0avgdata 9760maxresident)k
0inputs+0outputs (0major+2507minor)pagefaults 0swaps
```